### PR TITLE
Log document access

### PR DESCRIPTION
**Pull request by bengolder**
This is an updated version of https://github.com/codeforamerica/vita-min/pull/901

- it makes the migration more reliable
- it adds document access logging to the edit view

### CODE REVIEW

File: app/controllers/hub/documents_controller.rb [hunk 1]
```diff
@@ -17,12 +17,15 @@ def index
...
+      log_document_access!
       redirect_to transient_storage_url(@document.upload.blob)
     end
 
     def new; end
 
-    def edit; end
+    def edit
+      log_document_access!
+    end
```
Comment (bengolder) [inline app/controllers/hub/documents_controller.rb:28]: I added this in response to https://github.com/codeforamerica/vita-min/pull/901#issuecomment-792976981

File: db/migrate/20210312004640_add_record_to_access_log.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    change_column_null :access_logs, :client_id, true
+  end
+
+  def down
+    # remove the old columns
+    remove_reference :access_logs, :record, polymorphic: true
+    # we can't reset client_id to null false, it's not safe
+  end
+end
+
```
Comment (bengolder) [inline db/migrate/20210312004640_add_record_to_access_log.rb:20]: The previous migration:
https://github.com/codeforamerica/vita-min/pull/901/files#diff-43b5d75f9dc345eb9e4109db7ef08d244489f81ed86238e6dedadd30dd6f3014R1-R15

Once this is deployed, we can follow up with a migration that removes the client_id column

File: app/controllers/hub/documents_controller.rb [hunk 3]
```diff
@@ -17,12 +17,15 @@ def index
     end
 
     def show
+      log_document_access!
```
Comment (alaroia-cfa) [inline app/controllers/hub/documents_controller.rb:20]: Consider adjusting we get the URL so it always logs document by default.

File: app/controllers/hub/documents_controller.rb (hunk 1 above)
Comment (alaroia-cfa) [reply app/controllers/hub/documents_controller.rb:28]: This seems unnecessary; is peaceful I suppose to keep! The document URL doesn't actually get sent to any web browsers through this method.

File: app/controllers/hub/documents_controller.rb [hunk 4]
```diff
@@ -48,6 +51,17 @@ def update
 
     private
 
+    def log_document_access!
+      AccessLog.create!(
+        user: current_user,
+        record: @document,
+        created_at: DateTime.now,
```
Comment (alaroia-cfa) [inline app/controllers/hub/documents_controller.rb:58]: flawless

File: db/migrate/20210312004640_add_record_to_access_log.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class AddRecordToAccessLog < ActiveRecord::Migration[6.0]
+  def up
+    # add the new columns (does this make indices?)
```
Comment (alaroia-cfa) [inline db/migrate/20210312004640_add_record_to_access_log.rb:3]: Can toss `(...)` in comment

File: db/migrate/20210312004640_add_record_to_access_log.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    add_reference :access_logs, :record, polymorphic: true
+    # set the new columns
+    update "UPDATE access_logs SET record_type = 'Client', record_id = client_id"
+    # make the new columns non-nullable
+    change_column_null :access_logs, :record_id, false
```
Comment (alaroia-cfa) [inline db/migrate/20210312004640_add_record_to_access_log.rb:8]: Making this non-nullable is unsafe for the same reason as deleting a column.

File: db/migrate/20210312004640_add_record_to_access_log.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    change_column_null :access_logs, :record_type, false
+    # make the old column nullable
+    change_column_null :access_logs, :client_id, true
+  end
+
+  def down
+    # remove the old columns
+    remove_reference :access_logs, :record, polymorphic: true
```
Comment (alaroia-cfa) [inline db/migrate/20210312004640_add_record_to_access_log.rb:16]: One option if you want is to delete the AccessLogs that don't foreign-key to a Client. That's sorta weird so let's that not write that code and hope we never need to do anything like that.

**Review by alaroia-cfa — CHANGES_REQUESTED** [review]
Looks great overall! Please do change the migration so that we set `record_id` and `record_type` to non-null in the _next_ migration that removes `client_id`. Other than that, 🚢

**Comment by bengolder** [comment]
Requested changes have been added! Merging now.

**Pull request merged**